### DISCUSSION TITLE
Linting tool `trunk` seems to hang and it's probably related to aqua proxy

### DISCUSSION THREAD

**Discussion by sheldonhull** [aquaproj/aqua#2025] [Deprecated (Support Request)]
### aqua version

`aqua version 2.7.0 (12ad6be65e139f389975d23b34f735a0ff5a59d0)`

### Environment

* OS: macOS
* CPU Architecture: amd65
 

### Overview

[trunk.io](https://trunk.io) is a great linting toolchain for pre-commit and pre-push. 

I've run into some odd issues that seem related to being stuck when trying to invoke certain commands using the system Go install (aka aqua managed) vs the "hermetically sealed" Go that trunk would install.

This adds more setup time and I prefer to simplify the management by using Aqua for Go + golangci-lint instead of double installing via trunk.

### How to reproduce

```shell
mkdir `trunk-test` && cd trunk-test
git init
echo 'test readme.md' > README.md
go mod init dev.local && go mod tidy && git add go.mod && git add go.sum
aqua init && git add aqua.yaml
git add README.md && git commit -m"chore: initial commit"
```

Now run: 

- `trunk init` and it should setup latest in the repo.
- Should detect Go tooling, and have golangci-lint installed.
- Run `aqua generate golang/go@go1.20.4 && aqua generate -i golangci/golangci-lint@v1.52.2 && aqua i`
- `trunk init` and `trunk install` should work without hanging.

Since we have the tools installed now, open `trunk.yaml`. 

Find the golangci-lint@vx.x.x and replace the version with `SYSTEM` string so it tries to use whatever our profile has in PATH.
Then do the same for the runtimes that mention `go@1.20.3` or whatever and put `go@SYSTEM`.

I get go tooling hanging without any resolution and `htop --filter=go` shows `go install ....` just stuck or looping (didn't confirm process id changed), and tool indefinitely won't work. 

### Expected behaviour

I should be able to use tooling managed by Go with trunk.

### Actual behaviour

I can't tell if it's a trunk issue (they are stumped right now), or what feels more likely is some type of issue with the proxyed request coming through not from me, but from another process invoking (and I confirmed it's not doing a subshell, should be respecting my session). 

The only thing that makes me suspicious that trunk invocation isn't calling in the context of my current session is the caller seems to be bash instead of trunk... 

### Important Factoids

This is a bit more complex of an issue, but I want to bring it up as the _only_ two issues I think I've seen based on the incredible design of aqua are:

- Go/gopls having issues in tooling like VSCode due to proxied request.
- Changes recently on how tooling was executed in linux and macOS.
- This second issue also might be related to the aqua proxy.

I think aqua proxying requests is an ingenious design, so I think working through this and having some good answers for cases that can't be fixed or what to look for will help further increase the ease of adoption when bumps come along.

Please don't consider this urgent, but I'd like to do this as a investigation together to improve aqua further. 

### Reference

*

**Comment by suzuki-shunsuke** [comment]
I'm not familiar with trunk, and I don't understand what is trunk and how trunk is related to aqua.

I install trunk and try to reproduce the issue on my macOS (darwin/amd64).

https://docs.trunk.io/docs/install#with-homebrew-macos-only

```console
$ brew install trunk-io
$ trunk --version
1.10.0

$ aqua -v
aqua version 2.7.0 (12ad6be65e139f389975d23b34f735a0ff5a59d0)
```

```console
$ which go
/Users/shunsuke-suzuki/.local/share/aquaproj-aqua/bin/go

$ which golangci-lint
/Users/shunsuke-suzuki/.local/share/aquaproj-aqua/bin/golangci-lint

$ go version
go version go1.20.4 darwin/amd64

$ golangci-lint version
golangci-lint has version 1.52.2 built with go1.20.2 from da04413a on 2023-03-25T18:11:28Z
```

```sh
mkdir trunk-test
cd trunk-test
git init
echo 'test readme.md' > README.md

go mod init dev.local
go mod tidy

aqua init
git add .
git commit -m "chore: initial commit"
```

```sh
aqua g -i golang/go@go1.20.4 golangci/golangci-lint@v1.52.2
aqua i
```

```console
$ ls -A
.git  README.md  aqua.yaml  go.mod

$ git status
On branch main
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   aqua.yaml
```

```console
$ trunk init
trunk init
Initializing 100% [=================================================================================================================================================>]  9/9  1.4s 
Formatting Configs 100% [===========================================================================================================================================>]  7/7  0.5s 
                                                                                                                                                                                  
✔ 5 linters were enabled (.trunk/trunk.yaml)

  git-diff-check (5 files)
  gitleaks 8.16.3 (5 files)
  markdownlint 0.34.0 (1 markdown file) (created .markdownlint.yaml)
  prettier 2.8.8 (1 markdown, 2 yaml files)
  yamllint 1.32.0 (2 yaml files) (created .yamllint.yaml)

→ Would you like Trunk to manage your git hooks? (Y/n): 
→ Run 'trunk fmt' whenever you run 'git commit'? (Y/n): 
→ Run 'trunk check' whenever you run 'git push'? (Y/n): 

✔ 3 actions were enabled (.trunk/trunk.yaml)

  trunk-announce 
  trunk-check-pre-push 
  trunk-fmt-pre-commit 

✔ Trunk is now managing your git hooks (https://docs.trunk.io for more info)

→ Would you like to sample current linter issues in your repo? (Y/n): 
Running trunk check --sample 5
Checking 100% [===================================================================================================================================================>]  33/33  1.0s 
  AUTOFIXES                                                                                                                                                                       

aqua.yaml
 1:1  high  Incorrect formatting  

   8 | #   - all
   9 | registries:
  10 | - type: standard
  11 |   ref: v4.9.0 # renovate: depName=aquaproj/aqua-registry
     |   - type: standard
     |     ref: v4.9.0 # renovate: depName=aquaproj/aqua-registry
  12 | packages:
  13 | - name: golang/go@go1.20.4
  14 | - name: golangci/golangci-lint@v1.52.2
     |   - name: golang/go@go1.20.4
     |   - name: golangci/golangci-lint@v1.52.2


  ISSUES  

README.md:1:0
 1:0  low  First line in a file should be a top-level heading  markdownlint/MD041

aqua.yaml
 -:-  fmt  Incorrect formatting, autoformat by running 'trunk fmt'  prettier

Checked 7 files
✖ 2 issues (1 auto-fixable)

Next Steps

 1. Read documentation
    Our documentation can be found at https://docs.trunk.io

 2. Get help and give feedback
    Join the Trunk community at https://slack.trunk.io
```

```console
$ trunk install
trunk install
Installing 100% [===================================================================================================================================================>]  7/7  0.2s 
✔ Done
```

https://docs.trunk.io/docs/reference-trunk-yaml

```console
$ cat .trunk/trunk.yaml
version: 0.1
cli:
  version: 1.10.0
plugins:
  sources:
    - id: trunk
      ref: v0.0.17
      uri: https://github.com/trunk-io/plugins
lint:
  enabled:
    - git-diff-check
    - gitleaks@8.16.3
    - markdownlint@0.34.0
    - prettier@2.8.8
    - yamllint@1.32.0
runtimes:
  enabled:
    - go@1.19.5
    - node@18.12.1
    - python@3.10.8
actions:
  enabled:
    - trunk-announce
    - trunk-check-pre-push
    - trunk-fmt-pre-commit
    - trunk-upgrade-available
```

> Since we have the tools installed now, open trunk.yaml.
> 
> Find the [golangci-lint@vx.x.x](mailto:golangci-lint@vx.x.x) and replace the version with SYSTEM string so it tries to use whatever our profile has in PATH.

I can't find golangci-lint in my .trunk/trunk.yaml 🤔

**Reply by suzuki-shunsuke** [reply]
https://docs.trunk.io/docs/check-supported-linters#golangci-lint

> Make sure your go version in go.mod matches Trunk's go runtime version.
> At the time you writing, Trunk's default go runtime version is 1.18.3.
> You can find out what it is via trunk print-config, and look for the runtime section, and you can override the default version in your trunk.yaml via:

**Reply by suzuki-shunsuke** [reply]
I fixed .trunk/trunk.yaml.

```console
$ cat .trunk/trunk.yaml 
version: 0.1
cli:
  version: 1.10.0
plugins:
  sources:
    - id: trunk
      ref: v0.0.17
      uri: https://github.com/trunk-io/plugins
lint:
  enabled:
    - git-diff-check
    - gitleaks@8.16.3
    - markdownlint@0.34.0
    - prettier@2.8.8
    - yamllint@1.32.0
    - golangci-lint@SYSTEM
runtimes:
  enabled:
    - go@SYSTEM
    - node@18.12.1
    - python@3.10.8
actions:
  enabled:
    - trunk-announce
    - trunk-check-pre-push
    - trunk-fmt-pre-commit
    - trunk-upgrade-available
```

Then how can we reproduce the issue?

**Reply by sheldonhull** [reply]
> I fixed .trunk/trunk.yaml.
> 
> 
> 
> ```console
> 
> $ cat .trunk/trunk.yaml 
> 
> version: 0.1
> 
> cli:
> 
>   version: 1.10.0
> 
> plugins:
> 
>   sources:
> 
>     - id: trunk
> 
>       ref: v0.0.17
> 
>       uri: https://github.com/trunk-io/plugins
> 
> lint:
> 
>   enabled:
> 
>     - git-diff-check
> 
>     - gitleaks@8.16.3
> 
>     - markdownlint@0.34.0
> 
>     - prettier@2.8.8
> 
>     - yamllint@1.32.0
> 
>     - golangci-lint@SYSTEM
> 
> runtimes:
> 
>   enabled:
> 
>     - go@SYSTEM
> 
>     - node@18.12.1
> 
>     - python@3.10.8
> 
> actions:
> 
>   enabled:
> 
>     - trunk-announce
> 
>     - trunk-check-pre-push
> 
>     - trunk-fmt-pre-commit
> 
>     - trunk-upgrade-available
> 
> ```
> 
> 
> 
> Then how can we reproduce the issue?

Thank you! Now that you added system to go and golangcilint does trunk install or trunk check --all work? Maybe add a shell script and make sure shfmt linter is added by running trunk upgrade? 

It hangs for me only when invoking with aqua managed go. 

I'm also using go as shell script calling in /usr/local/bin/go to avoid the issues with VSCode crashing.

**Reply by suzuki-shunsuke** [reply]
go is managed by aqua but `trunk check --all` outputs `go not found on the system`. 🤔

```console
$ trunk install
Installing 0% [>                                                                                                                                                     ]  0/0  0.2s 
✔ Done                                                                                                                                                                            

$ trunk check --all
Checking 50% [===========================================================================>                                                                           ]  1/2  0.0s 
✖ go not found on the system                                                                                                                                                      

$ which go
/Users/shunsuke-suzuki/.local/share/aquaproj-aqua/bin/go

$ go version
go version go1.20.4 darwin/amd64
```

**Reply by suzuki-shunsuke** [reply]
> Maybe add a shell script and make sure shfmt linter is added by running trunk upgrade?

I added `main.sh` and ran `trunk upgrade`, but shfmt wasn't added to trunk.yaml.

```console
$ cat main.sh
#!/usr/bin/env bash


  echo hello

$ trunk upgrade
Checking for upgrades 100% [========================================================================================================================================>]  4/4  1.5s 
                                                                                                                                                                                  
Already up to date

$ cat .trunk/trunk.yaml
version: 0.1
cli:
  version: 1.10.0
plugins:
  sources:
    - id: trunk
      ref: v0.0.17
      uri: https://github.com/trunk-io/plugins
lint:
  enabled:
    - git-diff-check
    - gitleaks@8.16.3
    - markdownlint@0.34.0
    - prettier@2.8.8
    - yamllint@1.32.0
    - golangci-lint@SYSTEM
runtimes:
  enabled:
    - go@SYSTEM
    - node@18.12.1
    - python@3.10.8
actions:
  enabled:
    - trunk-announce
    - trunk-check-pre-push
    - trunk-fmt-pre-commit
    - trunk-upgrade-available
```

**Reply by sheldonhull** [reply]
Sorry, try this: `trunk check enable shfmt`. That should add. Then see if it hangs for you?
The go not found on system, maybe that's because you don't have `/usr/local/bin/go` as the shell script running go via aqua? If you look at the process running in `htop --filter=go` I bet it's running via "bash" and not taking the AQUA_ROOT_DIR into account. That's my guess. 


Like I said this take your time as we work through this. Just trying to iron out the edge cases where other more complex dev tools aren't playing nice. 

Thanks for the quick follow-up as usual 👍
